File: py/test_money.py
import unittest
from money import Money
from portfolio import Portfilo

class TestMoney(unittest.TestCase):
    def testMultiplication(self):
        tenEuros = Money(10,"EUR")
        twentyEuros = Money(20, "EUR")
        self.assertEqual(twentyEuros, tenEuros.times(2))
    def testDivision(self):
        origianlMoney = Money(2500.5, "KRW")
        actualMoneyAfterDivision = origianlMoney.divide(2.5)
        expectedMoneyAfterDivision = Money(1000.2, "KRW")
        self.assertEqual(expectedMoneyAfterDivision,
                         actualMoneyAfterDivision)
    def testAddition(self):
        fiveDollars = Money(5, "USD")
        tenDollars = Money(10, "USD")
        fifteenDollars = Money(15, "USD")
        portfilo = Portfilo()
        portfilo.add(fiveDollars, tenDollars)
        self.assertEqual(fifteenDollars, portfilo.evaluate("USD"))

    def testAdditionOfDollarsAndEuros(self):
        fiveDollars = Money(5, "USD")
        tenEuros = Money(10, "EUR")
        portfilo = Portfilo()
        portfilo.add(fiveDollars, tenEuros)
        expectedValue = Money(17, "USD")
        actualValue = portfilo.evaluate("USD")
        self.assertEqual(
            expectedValue, actualValue, "%s != %s" %(expectedValue, actualValue)
        )

    def testAdditionOfDollarsAndWons(self):
        oneDollar = Money(1, "USD")
        elevenHundredWon = Money(1100, "KRW")
        portfilo = Portfilo()
        portfilo.add(oneDollar, elevenHundredWon)
        expectedValue = Money(2200, "KRW")
        actualValue = portfilo.evaluate("KRW")
        self.assertEqual(
            expectedValue, actualValue, "%s != %s" %(expectedValue, actualValue)
        )

if __name__ == '__main__':
    unittest.main()

File: js/test_money.js
const assert = require('assert');
const Money = require('./Money');
const Portfolio = require('./Portfoilo');
const exp = require('constants');
const Bank = require('./Bank');

class MoneyTest{
    setUp(){
        this.bank = new Bank();
        this.bank.addExchangeRate("EUR", "USD", 1.2);
        this.bank.addExchangeRate("USD", "KRW", 1100);
    }

    testMultiplication()
    {
        let tenEuros = new Money(10, "EUR")
        let twentyEuros = new Money(20, "EUR")
        assert.deepStrictEqual(tenEuros.times(2), twentyEuros)        
    }
    testDivision()
    {
        let originalMoney = new Money(2500.5, "KRW")
        let actualMoneyAfterDivision = originalMoney.divide(2.5)
        let expectedMoneyAfterDivision = new Money(1000.2, "KRW")
        assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision)        
    }

    testAddition()
    {
        let fiveDollars = new Money(5, "USD");
        let tenDollars = new Money(10, "USD");
        let fifteenDollars = new Money(15, "USD");
        let portfolio = new Portfolio();
        portfolio.add(fiveDollars, tenDollars);
        assert.deepStrictEqual(portfolio.evaluate(this.bank, "USD"), fifteenDollars);
    }

    testAdditionOfDollarsAndEuros(){
        let fiveDollars = new Money(5, "USD");
        let tenEuros = new Money(10, "EUR");
        let portfolio = new Portfolio();

        portfolio.add(fiveDollars, tenEuros);

        let expectedValue = new Money(17, "USD");
        assert.deepStrictEqual(portfolio.evaluate(this.bank, "USD"), expectedValue);
    }

    testAdditionOfDollarsAndWons(){
        let oneDollar = new Money(1, "USD");
        let elevenHundredWon = new Money(1100, "KRW");
        let portfolio = new Portfolio();

        portfolio.add(oneDollar, elevenHundredWon);
        let expectedValue = new Money(2200, "KRW");
        assert.deepStrictEqual(portfolio.evaluate(this.bank, "KRW"), expectedValue);
    }

    testAdditionWithMultipleMissingExchangeRates(){
        let oneDollar = new Money(1, "USD")
        let oneEuro = new Money(1, "EUR");
        let oneWon = new Money(1, "KRW");
        let portfolio = new Portfolio();
        portfolio.add(oneDollar, oneEuro, oneWon);
        let expectedError = new Error(
            "Missing exchange rate(s):[USD->Kalganid,EUR->Kalganid,KRW->Kalganid]");
        assert.throws(() => portfolio.evaluate(this.bank, "Kalganid"), expectedError);
    }

    testConversion(){
        let tenEuros = new Money(10, "EUR")
        assert.deepStrictEqual(
            this.bank.convert(tenEuros, "USD"), new Money(12, "USD")
        );

        this.bank.addExchangeRate("EUR", "USD", 1.3);
        assert.deepStrictEqual(
            this.bank.convert(tenEuros, "USD"), new Money(13, "USD")
        );
    }

    // testWhatIsTheConversionRateFromEURToUSD(){
    //     let tenEuros = new Money(10, "EUR")
    //     assert.deepStrictEqual(
    //         this.bank.convert(tenEuros, "USD"), new Money(12, "USD")
    //     );
    // }

    getAllTestMethods(){
        let moneyProtype = MoneyTest.prototype;
        let allProps = Object.getOwnPropertyNames(moneyProtype);
        let testMethods = allProps.filter(p => {
            return typeof moneyProtype[p] === 'function' && p.startsWith("test");
        });
        return testMethods;
    }

    runAlltests()
    {
        let testMethods = this.getAllTestMethods();
        testMethods.forEach(m =>{
            console.log("Running: %s()", m);
            let method = Reflect.get(this, m);
            try{
                this.setUp();
                Reflect.apply(method, this, []);
            }catch(e){
                if(e instanceof assert.AssertionError){
                    console.log(e);
                }else{
                    throw e;
                }
            }
        });
    }
}

new MoneyTest().runAlltests();